src/components/input/SelectNext.tsx

```diff
@@ -0,0 +1,195 @@
+  const wrapperRef = useRef<HTMLDivElement | null>(null);
+  const listboxRef = useRef<HTMLDivElement | null>(null);
+  const listboxId = useId();
+  const buttonRef = useSyncedRef(elementRef);
+  const buttonId = useId();
+
+  const selectValue = useCallback(
+    (newValue: unknown) => {
+      onChange(newValue as T);
+      closeListbox();
+    },
+    [closeListbox, onChange],
+  );
+
+  // When clicking away or pressing `Esc`, close the listbox
+  useClickAway(wrapperRef, closeListbox);
+  useKeyPress(['Escape'], closeListbox);
+
+  // Vertical arrow key for options in the listbox
+  useArrowKeyNavigation(wrapperRef, { horizontal: false });
+
+  useLayoutEffect(() => {
+    if (!isListboxOpen) {
+      // Focus button after closing listbox
+      buttonRef.current!.focus();
+      // Reset shouldDropUp so that it does not affect calculations next time
+      // it opens
+      setShouldListboxDropUp(false);
+    } else {
+      const viewportHeight = window.innerHeight;
+      const { top: buttonDistanceToTop, bottom: buttonBottom } =
+        buttonRef.current!.getBoundingClientRect();
+      const buttonDistanceToBottom = viewportHeight - buttonBottom;
+      const { bottom: listboxBottom } =
+        listboxRef.current!.getBoundingClientRect();
+      const listboxDistanceToBottom = viewportHeight - listboxBottom;
+
+      // The listbox should drop up only if there's not enough space below it for
+      // the listbox max height, and there's also more space above
+      setShouldListboxDropUp(
+        listboxDistanceToBottom < 0 &&
+          buttonDistanceToTop > buttonDistanceToBottom,
+      );
+    }
+  }, [buttonRef, isListboxOpen]);
+
+  return (
+    <div className="relative" ref={wrapperRef}>
+      <Button
+        id={buttonId}
+        variant="custom"
+        classes={classnames(
+          'w-full flex border',
+          'bg-grey-0 disabled:bg-grey-1 disabled:text-grey-6',
+          classes,
+        )}
+        expanded={isListboxOpen}
+        pressed={isListboxOpen}
+        disabled={disabled}
+        aria-haspopup="listbox"
+        aria-controls={listboxId}
+        elementRef={buttonRef}
+        onClick={() => setIsListboxOpen(prev => !prev)}
+        onKeyDown={e => {
+          if (e.key === 'ArrowDown' && !isListboxOpen) {
+            setIsListboxOpen(true);
+          }
+        }}
+        data-testid="select-toggle-button"
+      >
+        {label}
+        <div className="grow" />
+        {isListboxOpen ? <MenuCollapseIcon /> : <MenuExpandIcon />}
+      </Button>
+      <SelectContext.Provider value={{ selectValue, value }}>
+        <div
+          className={classnames(
+            'absolute z-5 w-full max-h-80 overflow-y-auto',
+            'rounded border bg-white shadow hover:shadow-md focus-within:shadow-md',
+            {
+              'top-full mt-1': !shouldListboxDropUp,
+              'bottom-full mb-1': shouldListboxDropUp,
+              hidden: !isListboxOpen,
```

Comment:
There are two main reasons to hide the listbox rather than unmounting it:

* Screen readers can identify the toggle button as a listbox handler (see `aria-haspopup="listbox"` and `aria-labelledby={buttonId}` on the listbox itself)
* It allows to properly calculate the height the listbox is going to need, in order to determine if it should drop down or drop up.

Reply:
This makes sense. Perhaps add a short version of this code comment in the code itself.